keps/sig-architecture/3136-beta-apis-off-by-default/README.md

```diff
@@ -0,0 +1,293 @@
+-->
+
+[kubernetes.io]: https://kubernetes.io/
+[kubernetes/enhancements]: https://git.k8s.io/enhancements
+[kubernetes/kubernetes]: https://git.k8s.io/kubernetes
+[kubernetes/website]: https://git.k8s.io/website
+
+## Summary
+
+New beta APIs will not be enabled in clusters by default.
+Existing beta APIs and new versions of existing beta APIs, will continue to be enabled by default: 
+if v1beta.some.group is currently enabled by default and we create v1beta2.some.group, v1beta2.some.group will still be enabled by default.
+
+## Motivation
+
+Beta APIs are not considered stable and reliance upon APIs in this state leads to exposure to bugs,
+guaranteed migration pain for users when the APIs move to stable, and the risk that dependencies will
+grow around unfinished APIs.
+Enabling beta APIs by default, exacerbates these problems by making them on in nearly every cluster.
+We observed these problems as we removed long-standing beta APIs and the PRR survey tells us that over
+90% of production clusters leave these APIs enabled.
+Unsuitability for production use is documented at https://kubernetes.io/docs/reference/using-api/#api-versioning 
+("The software is not recommended for production uses"), but defaulting on means they are present in nearly every 
+production cluster.
+By disabling beta APIs by default, a cluster-admin can opt-in for specific APIs without having every
+incomplete API present in the cluster.
+
+### Goals
+
+1. Disable new beta APIs by default.
+2. Continue enabling existing beta APIs and new version of existing beta APIs by default:
+   if v1beta.some.group is currently enabled by default and we create v1beta2.some.group, v1beta2.some.group will still be enabled by default.
+3. Allow enabling specific resources in beta.  Enable coolnewjobtype.v1beta1.batch.k8s.io without enabling other-neat-job.v1beta1.batch.k8s.io
+
+### Non-Goals
+
+1. Change featuregate defaults.
+
+## Proposal
+
+New beta APIs will be placed into the `DisableVersions` stanza instead of the `EnableVersions` stanza (see [DefaultAPIResourceConfigSource](https://github.com/kubernetes/kubernetes/blob/0669da445fa8c1ae07c15c0827f0e83da11cbe58/pkg/controlplane/instance.go#L643)).
+The `--runtime-config` flag will be extended to allow `group/version/resource=true`, to enable specific resources.
+To enable a beta API, a cluster-admin will have to add the appropriate `--runtime-config` flags.
+
+### User Stories (Optional)
+
+#### Story 1
+
+As a cluster-admin I want to enable the coolnewjobtype.v1beta1.batch.k8s.io API in my cluster.
+
+To do this I call `kube-apiserver --runtime-config=batch.k8s.io/v1beta1/coolnewjobtype`.
+
+#### Story 2
+
+As a cluster-admin I want to enable all beta APIs as in past releases.
+
+To do this I call `kube-apiserver --runtime-config=api/beta=true`.
+This already exists and will continue to function.
+
+
```

Comment:
Another risk: since this greatly diminishes the impact of making something beta, we very well may end up with much worse betas. Beta is the new alpha, GA 1.0 is the new beta, GA 2.0 is the new GA 1.0. Only difference is that the new alpha is a lot more work / overhead than the old alpha.

Reply:
I've tried to address what I think this is getting at in another risk.  I think we have to trust sigs to continue treating their beta APIs as, "nearly there".

Reply:
I think that we should distinguish two things:
- quality of the API itself
- quality of the implementation of the features backing this API

Regarding the first one, I believe in our API approval process and I'm not that worried about this.

Regarding the second one - PRR is supposed to protect us from Beta becoming a new Alpha. Of course this only covers some aspects, there is also a question about usefulness/completeness of the features, but I'm actually challenging some things in these are during PRR reviews too (and I've seen other PRR approvers doing that too), so I'm not that worried about this.

So in other words - I don't think the risk of degrading Beta quality is very high.
I think the risk is that the quality of GA will degrade (which I see you mentioned explicitly below).

Thinking more about this, I'm wondering if we don't want to try mitigating it by explicit addition/extensions to our PRR and/or KEP template itself, that will be more about "user feedback". In other words, we keep a road to beta as is, but in order to reach GA, you no longer need to just say "no complaints from users" (which is a common practice now), but we add an explicit section/questions/whatever that is focused on user feedback that we got in Beta [to mitigate the risk that we won't reach Beta with no/very weak feedback.]
@johnbelamaric @deads2k @ehashman - thoughts from PRR side?

Reply:
I'm not flat out opposed to that idea, wojtek. But it seems like a little bit of scope creep for PRR, which was intended more to review from an observability / supportability / scalability / operational safety point of view. The functional completeness point of view is somewhat different. I think it's fine to push back on obvious gaps but making it a formal part of the process I am not sure about. I would hope that the SIG leadership would be taking the functional completeness aspect into account, before proposing an enhancement for graduation.

What sort of criteria would you want to see in such a KEP question? Specific comments / requests and what was done about them? Number of confirmed users? Instances of use at scale? Some sort of survey? Mailing list discussions?

Reply:
@johnbelamaric - sorry, I think I wasn't clear enough with I was proposing.
So let me rephrase what I was thinking about (that's more of brainstorm than a real proposal now):

1) I'm definitely not proposing adding anything about feature completeness to PRR - that's not the place for it. Maybe in KEP template, but even that seems far from obvious to me. I agree, it's more on SIG leadership to ensure the feature we're pushing makes sense.

2) What I was suggesting was to ensure that a feature that is going to GA was actually validated in some production (or production-ish) environment really. To me this is something we need to validate quality, which PRR to me is also about. With beta APIs being off by-default, in theory it can happen that you will be able to get to GA without any real usage, even if it was sitting in beta for 3 releases (it isn't the case now). So adding some question about "was this validated for real production usacases", sounds like a valid addition to me.
I don't have a specific question on my mind - I don't want to make it too heavy. But maybe requiring something like brief description of production usage validating the usage (e.g. what version, env (e.g. what (cloud-)provider), scale of cluster (not necessary about size of cluster, it can be other dimension to more relevant to this API), and optionally list of found issues will not be that hard and will actually prove that we really did validate that?

Reply:
added to criteria here: https://github.com/kubernetes/enhancements/pull/3137/files#diff-0ad9f743dbfdf3932906414acf0a79e074af723ad2ffb8e0dab3429f0b664beeR261